Embed: IPC (ready signaling, navigation) (#2314)
quiltdata/quilt: docs/Catalog/Embed.md

```js
import * as React from 'react'

const EMBED_ORIGIN = 'https://my-quilt-catalog'
const PARENT_ORIGIN = window.location.origin
const EVENT_SOURCE = 'quilt-embed'

const mkNonce = () => `${Math.random()}`.slice(2)

function Embed() {
  const iframeRef = React.useRef(null)

  const [nonce, setNonce] = React.useState(mkNonce)

  // nonce is used to identify "our" embed instance and make sure we're receiving messages from the same instance
  // origin must be sent as a query parameter to enable cross-origin message passing from embed to the parent
  const src = `${EMBED_ORIGIN}/__embed?nonce=${nonce}&origin=${encodeURIComponent(PARENT_ORIGIN)}`

  const postMessage = React.useCallback(
    // function for sending messages to the embed (it will only handle messages sent by the window that opened it aka parent)
    (msg) => {
      if (!iframeRef.current) return
      // origin must be sent as a second parameter to enable cross-origin message passing
      iframeRef.current.contentWindow.postMessage(msg, EMBED_ORIGIN)
    },
    [iframeRef],
  )

  const initialize = React.useCallback(() => {
    postMessage({

>       type: 'init',

Reviewer: Maybe highlight in comments which parameters are required and which are optional

Author: i've added this info to the API reference section

> The Quilt catalog has a special (limited) S3 browser UI that can be embedded into any web app using an iframe -- "Embed" further on.

Reviewer 2: ```suggestion
The Quilt catalog has limited-feature S3 browser called Embed, an embeddable iframe.
```

This embeddable browser is served off the main catalog server under `/__embed` route.

![](../imgs/embed-example.png)

## Usage

> Here's an example code written in plain JavaScript + React outlining the basic use cases:

Reviewer 2: ```suggestion
Here's a sample code snippet in plain JavaScript + React that outlines the basic use cases:
```

    />
  )
}
```

## API reference

### URL and query parameters

The Embed is served off the main catalog server under `/__embed` route
which takes two optional query parameters:

`nonce` (any unique string) is used to identify "our" Embed instance and make sure we're receiving messages from that same instance.

`origin` must be sent to enable cross-origin message passing from Embed to the parent.

### Commands to Embed

Embed accepts commands from the parent via [`postMessage` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).
Command message data format:

```ts
interface Command {
  type: string
  // ...parameters
}
```

Available commands are listed below.

#### `init`

Initialize the Embed.
This command must be sent after the Embed is ready
(see `ready` message reference for details).

```ts
interface InitCommand extends Command {
  type: 'init'

  // Bucket name, e.g. 'my-bucket'
  bucket?: string
  // Path to object or prefix in the given bucket, e.g. 'some-prefix/some-object.csv'
  path?: string

  // Initial route Embed will navigate to, e.g. '/b/my-bucket/tree/some/path',
  // takes precedence over bucket / path
  route?: string

>   // Embed accepts any credentials supported by the Quilt authentication endpoint,

Reviewer 2: Link to list of support SSO providers https://docs.quiltdata.com/advanced-usage/technical-reference#single-sign-on-sso

  // for details.
  theme?: MUI.ThemeOptions // https://github.com/mui-org/material-ui/blob/v4.12.3/packages/material-ui/src/styles/createTheme.d.ts#L15

  // Some aspects of the UI can be overriden:
  overrides?: {
    // This prop is responsible for customizing the display and behaviour
    // of the "link" button in the object revision list menu
    s3ObjectLink?: {
      // Title of the link element
      title?: string
      // Link [template](https://lodash.com/docs/4.17.15#template).
      // Template context:
      //   url: string -- url / route of the object version in the context of the Embed
      //   s3HttpsUri: string -- HTTPS URI of the object version, e.g. https://my-bucket.s3.amazonaws.com/${key}?versionId=${version} (with properly encoded key)
      //   bucket: string -- current bucket
      //   key: string -- key of the browsed object
      //   version: string -- object version id
      // Example: 'https://my-app/s3-browser?route=<%= encodeURIComponent(url) %>'
      href?: string
      // Notification shown after copying href to the clipboard (if `emit` is not set to "override")
      notification?: string

>       // When set "notify" or "override", Embed will send "s3ObjectLink" message.

Reviewer 2: ```suggestion
      // When set "notify" or "override", Embed will send the "s3ObjectLink" message.
```
There should be alist of message types, even if it has only one item for now.

Author: other messages are not affected, it's only for `s3ObjectLink`

Sent when an error occurs.

```ts
interface ErrorMessage extends Message {
  type: 'error'
  message: string
  credentials?: object // for authentication error
  init?: object // for initialization error
  data?: object // for navigation error
}
```

#### `navigate`

Sent when the Embed navigates to a new route.
Useful for syncing Embed state with the parent app state.

```ts
interface NavigateMessage extends Message {
  type: 'navigate'
  route: string
  action: 'PUSH' | 'POP' | 'REPLACE'
}
```

#### `s3ObjectLink`

Enabled only when `overrides.s3ObjectLink.emit` parameter is set during initialization.
Sent when the link button in the object version menu is clicked.

![](../imgs/embed-object-link.png)

```ts
interface S3ObjectLinkMessage extends Message {
  type: 's3ObjectLink'
  // URL aka route of the object version in the context of the Embed
  url: string
  // HTTPS URI of the object version, e.g. https://my-bucket.s3.amazonaws.com/${key}?versionId=${version} (with properly encoded key)
  s3HttpsUri: string
  // Bucket the object resides in
  bucket: string
  // Object's key
  key: string
  // Object's version ID
  version: string
}
```

## Testing and debugging

> Catalog also has `/__embed-debug` route which serves a simple debug interface for the Embed:

Reviewer 2: ```suggestion
The catalog's `/__embed-debug` route is a simple driver for testing Embed:
```

![](../imgs/embed-debug.png)

> It's useful for trying different parameters and inspecting messages passed to/from the Embed.

Reviewer 2: ```suggestion
`__embed-debug` is useful for trying different parameters and inspecting messages passed to and from the Embed.
```

Its main components are:

> 1. Inputs for `init` parameters and button for sending `init` command.

Reviewer 2: ```suggestion
1. Inputs for `init` parameters and button for sending the `init` command.
```

> 2. "Navigate to" button and route input for sending `navigate` command.

Reviewer 2: ```suggestion
2. "Navigate to" button and route input for sending the `navigate` command.
```

  React.useEffect(() => {
    // subscribe to messages from Embed
    window.addEventListener('message', handleMessage)
    return () => {
      window.removeEventListener('message', handleMessage)
    }
  }, [handleMessage])

  return (
    <iframe
      ref={iframeRef}
      src={src}
      width="900"
      height="600"
      // other things like styling and stuff
    />
  )
}
```

## API reference

### URL and query parameters

The Embed is served off the main catalog server under `/__embed` route
which takes two optional query parameters:

`nonce` (any unique string) is used to identify "our" Embed instance and make sure we're receiving messages from that same instance.

`origin` must be sent to enable cross-origin message passing from Embed to the parent.

### Commands to Embed

Embed accepts commands from the parent via [`postMessage` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).
Command message data format:

```ts
interface Command {
  type: string
  // ...parameters
}
```

Available commands are listed below.

#### `init`

Initialize the Embed.
This command must be sent after the Embed is ready
(see `ready` message reference for details).

> Supported SSO providers are listed in the [Technical Reference](../Technical Reference.md#single-sign-on-sso).

Author: not sure why this doesn't work on gh

Author: works properly on gitbook tho
